flink-connectors/flink-connector-pubsub/src/main/java/org/apache/flink/streaming/connectors/pubsub/PubSubSink.java

```diff
@@ -0,0 +1,283 @@
+		Publisher.Builder builder = Publisher
+			.newBuilder(ProjectTopicName.of(projectName, topicName))
+			.setCredentialsProvider(serializableCredentialsProvider);
+
+		if (hostAndPort != null) {
+			managedChannel = ManagedChannelBuilder
+				.forTarget(hostAndPort)
+				.usePlaintext(true) // This is 'Ok' because this is ONLY used for testing.
+				.build();
+			channel = GrpcTransportChannel.newBuilder().setManagedChannel(managedChannel).build();
+			builder.setChannelProvider(FixedTransportChannelProvider.create(channel));
+		}
+
+		publisher = builder.build();
+	}
+
+	@Override
+	public void close() throws Exception {
+		super.close();
+		shutdownPublisher();
+		shutdownTransportChannel();
+		shutdownManagedChannel();
+	}
+
+	private void shutdownPublisher() {
+		try {
+			publisher.shutdown();
+		} catch (Exception e) {
+			LOG.info("Shutting down Publisher failed: " + e.getMessage());
+		}
+	}
+
+	private void shutdownTransportChannel() {
+		if (channel == null) {
+			return;
+		}
+		try {
+			channel.close();
+		} catch (Exception e) {
+			LOG.info("Shutting down TransportChannel failed: " + e.getMessage());
+		}
+	}
+
+	private void shutdownManagedChannel() {
+		if (managedChannel == null) {
+			return;
+		}
+		try {
+			managedChannel.shutdownNow();
+			managedChannel.awaitTermination(1000L, TimeUnit.MILLISECONDS);
+		} catch (Exception e) {
+			LOG.info("Shutting down ManagedChannel failed: " + e.getMessage());
+		}
+	}
+
+	@Override
+	public void invoke(IN message, SinkFunction.Context context) {
+		PubsubMessage pubsubMessage = PubsubMessage
+			.newBuilder()
+			.setData(ByteString.copyFrom(serializationSchema.serialize(message)))
+			.build();
+		publisher.publish(pubsubMessage);
+		publisher.publishAllOutstanding();
+	}
+
+	/**
+	 * Create a builder for a new PubSubSink.
+	 *
+	 * @param <IN> The generic of the type that is to be written into the sink.
+	 * @return a new PubSubSinkBuilder instance
+	 */
+	public static <IN> PubSubSinkBuilder<IN, ? extends PubSubSink<IN>, ? extends PubSubSinkBuilder<IN, ?, ?>> newBuilder() {
+		return new PubSubSinkBuilder<>(new PubSubSink<>());
+	}
+
+	/**
+	 * PubSubSinkBuilder to create a PubSubSink.
+	 *
+	 * @param <IN> Type of PubSubSink to create.
+	 */
+	@SuppressWarnings("unchecked")
+	public static class PubSubSinkBuilder<IN, PSS extends PubSubSink<IN>, BUILDER extends PubSubSinkBuilder<IN, PSS, BUILDER>> {
+		protected PSS sinkUnderConstruction;
+
+		private PubSubSinkBuilder(PSS sinkUnderConstruction) {
+			this.sinkUnderConstruction = sinkUnderConstruction;
+		}
+
+		/**
+		 * Set the credentials.
+		 * If this is not used then the credentials are picked up from the environment variables.
+		 *
+		 * @param credentials the Credentials needed to connect.
+		 * @return The current PubSubSinkBuilder instance
+		 */
+		public BUILDER withCredentials(Credentials credentials) {
+			sinkUnderConstruction.setSerializableCredentialsProvider(new SerializableCredentialsProvider(credentials));
```

Comment:
store the arguments in the builder instead, and pass all arguments to a constructor in `build()`. Validate the arguments either in the constructor or in `build()`. The fields can then all be final making the class more maintainable.

Reply:
Fixed